src/MESONT/angle_mesocnt.cpp

```diff
@@ -0,0 +1,392 @@
+    i2 = anglelist[n][1];
+    i3 = anglelist[n][2];
+    type = anglelist[n][3];
+
+    // 1st bond
+
+    delx1 = x[i1][0] - x[i2][0];
+    dely1 = x[i1][1] - x[i2][1];
+    delz1 = x[i1][2] - x[i2][2];
+
+    rsq1 = delx1 * delx1 + dely1 * dely1 + delz1 * delz1;
+    r1 = sqrt(rsq1);
+
+    // 2nd bond
+
+    delx2 = x[i3][0] - x[i2][0];
+    dely2 = x[i3][1] - x[i2][1];
+    delz2 = x[i3][2] - x[i2][2];
+
+    rsq2 = delx2 * delx2 + dely2 * dely2 + delz2 * delz2;
+    r2 = sqrt(rsq2);
+
+    // angle (cos and sin)
+
+    c = delx1 * delx2 + dely1 * dely2 + delz1 * delz2;
+    c /= r1 * r2;
+
+    if (c > 1.0) c = 1.0;
+    if (c < -1.0) c = -1.0;
+
+    s = sqrt(1.0 - c * c);
+    if (s < SMALL) s = SMALL;
+    s = 1.0 / s;
+
+    // force & energy
+
+    dtheta = acos(c) - theta0[type];
+
+    // harmonic bending
+    if (!buckling[type] || fabs(dtheta) < thetab[type]) {
+      tk = kh[type] * dtheta;
+      if (eflag) eangle = tk * dtheta;
+      a = -2.0 * tk * s;
+
+      buckled[i2] = 0;
+    }
+    // bending buckling
+    else {
+      if (eflag)
+        eangle = kb[type] * fabs(dtheta) + thetab[type] * (kh[type] * thetab[type] - kb[type]);
+      a = kb[type] * s;
+
+      buckled[i2] = 1;
+    }
+    a11 = a * c / rsq1;
+    a12 = -a / (r1 * r2);
+    a22 = a * c / rsq2;
+
+    f1[0] = a11 * delx1 + a12 * delx2;
+    f1[1] = a11 * dely1 + a12 * dely2;
+    f1[2] = a11 * delz1 + a12 * delz2;
+    f3[0] = a22 * delx2 + a12 * delx1;
+    f3[1] = a22 * dely2 + a12 * dely1;
+    f3[2] = a22 * delz2 + a12 * delz1;
+
+    // apply force to each of 3 atoms
+
+    if (newton_bond || i1 < nlocal) {
+      f[i1][0] += f1[0];
+      f[i1][1] += f1[1];
+      f[i1][2] += f1[2];
+    }
+
+    if (newton_bond || i2 < nlocal) {
+      f[i2][0] -= f1[0] + f3[0];
+      f[i2][1] -= f1[1] + f3[1];
+      f[i2][2] -= f1[2] + f3[2];
+    }
+
+    if (newton_bond || i3 < nlocal) {
+      f[i3][0] += f3[0];
+      f[i3][1] += f3[1];
+      f[i3][2] += f3[2];
+    }
+
+    if (evflag)
+      ev_tally(i1, i2, i3, nlocal, newton_bond, eangle, f1, f3, delx1, dely1, delz1, delx2, dely2,
+               delz2);
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+void AngleMesoCNT::allocate()
+{
+  allocated = 1;
+  const int np1 = atom->nangletypes + 1;
+
+  memory->create(buckling, np1, "angle:buckling");
+  memory->create(kh, np1, "angle:kh");
+  memory->create(kb, np1, "angle:kb");
+  memory->create(thetab, np1, "angle:thetab");
+  memory->create(theta0, np1, "angle:theta0");
+
+  memory->create(setflag, np1, "angle:setflag");
+  for (int i = 1; i < np1; i++) setflag[i] = 0;
+}
+
+/* ----------------------------------------------------------------------
+   set coeffs for one or more types
+------------------------------------------------------------------------- */
+
+void AngleMesoCNT::coeff(int narg, char **arg)
+{
+  if (narg < 1) error->all(FLERR, "Incorrect args for angle coefficients");
```

Comment:
We have recently changed how we prefer to handle errors. Rather than having terse and non-specific error messages in the source code, we want to include relevant details. However, they should be just enough to figure out the rest from the documentation rather than suggesting what to change already in the error message.

Some more detailed explanations and examples are here: https://lammps.org/workshops/Aug22/project1/ and recent pull requests: PR #3394, PR #3403, PR #3408